src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+                    }
+                }
+
+                while (currentSubShard.getEnd() < r.getStart()) {
+                    if (!readsSoFar.isEmpty()) {
+                        // ship this one.
+                        ret = shard.split(currentSubShard).withReads(readsSoFar);
+                        readsSoFar = new ArrayList<>();
+                    }
+                    // move to the next shard
+                    currentSubShard = subshards.get(++currentSubShardIndex);
+                }
+                // the header slows serialization too much
+                g.setHeader(null);
+                readsSoFar.add(g);
+
+                if (null!=ret) {
+                    // we have something to ship, and we're done handing the read we just got out of the query iterator.
+                    return ret;
+                }
+            }
+            // done reading, ship what we have
+            readerClosed = true;
+            query.close();
+            try {
+                if (null!=reader) {
+                    reader.close();
+                }
+            } catch (IOException x) {
+                throw new GATKException.ShouldNeverReachHereException("IOException when closing the BAM file reader for "+bam);
+            }
+            if (!readsSoFar.isEmpty()) {
+                return shard.split(currentSubShard).withReads(readsSoFar);
+            }
+            return null;
+        }
+
+        private boolean accept(GATKRead r, SimpleInterval region) {
+            if (r.isUnmapped()) {
+                return (!r.mateIsUnmapped() && r.getMateStart() >= region.getStart());
+            } else {
+                // mapped read.
+                // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                return r.getStart() >= region.getStart();
+            }
+        }
+
+    }
+
+
+
+    /**
+     * Given a list of shards and a list of variants,
+     * add each variant to every (shard+margin) that it overlaps.
+     *
+     * This happens immediately, at the caller.
+     */
+    public static ArrayList<ContextShard> fillVariants(List<SimpleInterval> shardedIntervals, List<Variant> variants, int margin) {
+        IntervalsSkipList<Variant> intervals = new IntervalsSkipList<>(variants);
+        ArrayList<ContextShard> ret = new ArrayList<>();
+        for (SimpleInterval s : shardedIntervals) {
+            int start = Math.max(s.getStart() - margin, 1);
+            int end = s.getEnd() + margin;
+            // here it's OK if end is past the contig's boundary, there just won't be any variant there.
+            SimpleInterval interval = new SimpleInterval(s.getContig(), start, end);
+            ret.add(new ContextShard(s).withVariants(intervals.getOverlapping(interval)));
```

Comment:
Yes. The shard's interval determines which reads it has. We get all the variants in the expanded interval to make sure that we have all the variants that overlap with our reads. ASCII graphics to the rescue:

````
   |------- s --------|
------s plus margin--------------------|
               |-- a read starting in s --|
                                     |--- a variant overlapping the read ---|
````
Since the read's length is less than margin, we know that by including all the variants that overlap with the expanded interval we are also including all the variants that overlap with all the reads in this shard.